### PR TITLE
[Do not merge] Travis CI updates

### PR DESCRIPTION
**Pull request by hornc**
Hi @cclauss, this builds upon your PR https://github.com/internetarchive/infogami/pull/5 and tries to get the Travis tests passing. It's not complete, and I'm not 100% sure I haven't broken other stuff, but this is what I've come up with in trying to investigate what is needed to bring Infogami towards Python3 so we can move forward with Open Library. Hopefully there is something useful in here.

I'm happy to raise this PR to another branch if you want to collaborate further and see if we can get a fully working version.

* I seem to have fixed the `E   TypeError: Can't instantiate abstract class Render with abstract methods __delitem__, __iter__, __len__, __setitem__` errors on your PR's test run by adding methods to `DictMixin` as suggested by the comments here: https://stackoverflow.com/questions/11165188/how-to-achieve-the-functionality-of-userdict-dictmixin-in-python-3

* This PR also adds a working db to CI and gets the tests a bit further

Currently the Python 2.7 tests are failing on the following main updating issues:

* `NameError: global name 'urllib' is not defined`, in many places, which is hopefully a straight forward compatibility / update fix. I just haven't dug into it yet. I think you sorted similar issues in OL?

* A number of `ImportError: No module named ClientForm` , which I have not investigated either, but again I'm hoping it is another import-the-module-correctly change.

* The other failures look more subtle and may have more to do with Infogami than python. I'd suggest dealing with those later...

### CODE REVIEW

**Comment by cclauss** [comment]
Flake8 still finds on malformed \N character escape __syntax error__ on Python 3.  This is similar to internetarchive/openlibrary#1468

https://travis-ci.org/internetarchive/infogami/jobs/495271374#L501-L503

$ __flake8 . --count --exit-zero --select=E999 --show-source --statistics__
```
./infogami/infobase/bulkupload.py:60:19: E999 SyntaxError: (unicode error) 'unicodeescape' codec can't decode bytes in position 0-1: malformed \N character escape
            return "\N"
                  ^
```

**Comment by cclauss** [comment]
https://github.com/infogami/infogami/issues

File: infogami/conftest.py [hunk 1]
```diff
@@ -1,5 +1,5 @@
 
-pytest_plugins = ["pytest_unittest"]
+# pytest_plugins = ["pytest_unittest"]
```
Comment (tfmorris) [inline infogami/conftest.py]: Can this just be deleted?

File: infogami/core/db.py [hunk 2]
```diff
@@ -4,13 +4,18 @@
...
+    basestring = (str, )
+
 def get_version(path, revision=None):
     return web.ctx.site.get(path, revision)
 
 @public
 def get_type(path):
     return get_version(path)
-    
+
```
Comment (tfmorris) [inline infogami/core/db.py]: All these whitespace changes make the diff very noisy

File: infogami/infobase/__init__.py [hunk 3]
```diff
@@ -1,10 +1,14 @@
 """
 Infobase.
 """
+from __future__ import print_function
+
+import logging
```
Comment (tfmorris) [inline infogami/infobase/__init__.py]: Appears to be unused

File: infogami/infobase/__init__.py [hunk 4]
```diff
@@ -1,10 +1,14 @@
...
 import sys
+
 import web
 
-import config, infobase, logger, logreader
+from infogami.infobase import config, infobase, logreader
```
Comment (tfmorris) [inline infogami/infobase/__init__.py]: logreader unused as well

File: infogami/infobase/tests/test_account.py [hunk 5]
```diff
@@ -15,15 +16,15 @@ def teardown_module(mod):
 class TestAccount:
     def setup_method(self, method):
         self.tx = db.transaction()
-    
+
```
Comment (tfmorris) [inline infogami/infobase/tests/test_account.py]: More whitespace noise

File: infogami/infobase/tests/pytest_wildcard.py [hunk 6]
```diff
@@ -26,7 +28,8 @@ def test_wildcard():
     assert 1 == wildcard
     assert ["foo", 1, 2] == [wildcard, 1, 2]
 
-def pytest_funcarg__wildcard(request):
+@pytest.fixture
```
Comment (tfmorris) [inline infogami/infobase/tests/pytest_wildcard.py]: Some of the test failures make me suspicious that this isn't working as intended, but need to dig in more

File: infogami/utils/template.py [hunk 7]
```diff
@@ -8,15 +8,21 @@
...
-import storage
+
+import web
+
+from infogami.utils import storage
+
 
 # There are some backward-incompatible changes in web.py 0.34 which makes Infogami fail. 
 # Monkey-patching web.py to fix that issue.
 if web.__version__ == "0.34":
```
Comment (tfmorris) [inline infogami/utils/template.py]: Not for this PR, but this might be something we want to clean up

File: infogami/utils/storage.py [hunk 8]
```diff
@@ -206,25 +206,39 @@ class DictPile(DictMixin):
     """
     def __init__(self, dicts=[]):
         self.dicts = dicts[:]
-        
+
+    def __len__(self):
+        return len(self.dicts)
```
Comment (tfmorris) [inline infogami/utils/storage.py]: I think len(self.keys()) is a more logical implementation.

File: infogami/utils/storage.py [hunk 9]
```diff
@@ -206,25 +206,39 @@ class DictPile(DictMixin):
...
+
+    def __iter__(self):
+        for i in self.dicts:
```
Comment (tfmorris) [inline infogami/utils/storage.py]: This should be an iterator over the merged dicts, but I'm not sure it's worth the aggravation to implement.

File: infogami/core/db.py (hunk 2 above)
Comment (cclauss) [reply infogami/core/db.py]: Sorry but today most editors and conversion tools (e.g. futurize, modernize) will automatically whack trailing whitespace in .py files.  It is a bit noisy now but should save aggravation on future PRs.

File: infogami/infobase/__init__.py (hunk 3 above)
Comment (cclauss) [reply infogami/infobase/__init__.py]: [flake8](http://flake8.pycqa.org) testing of https://github.com/internetarchive/infogami on Python 3.7.1

$ __flake8 . --select=F401 --statistics__
```
./test/bug_239238.py:11:1: F401 'infogami.infobase.infobase' imported but unused
./infogami/config.py:5:1: F401 'web' imported but unused
./infogami/plugins/links/code.py:9:1: F401 'web' imported but unused
./infogami/plugins/links/code.py:15:1: F401 'view' imported but unused
./infogami/plugins/i18n/code.py:8:1: F401 'infogami.config' imported but unused
./infogami/plugins/i18n/code.py:10:1: F401 'infogami.utils.context.context' imported but unused
./infogami/plugins/i18n/code.py:11:1: F401 'infogami.utils.view.public' imported but unused
./infogami/plugins/i18n/code.py:12:1: F401 'infogami.utils.template.render' imported but unused
./infogami/plugins/i18n/db.py:2:1: F401 'infogami.infobase.client' imported but unused
./infogami/infobase/logger.py:13:1: F401 'time' imported but unused
./infogami/infobase/logger.py:79:9: F401 'web' imported but unused
./infogami/infobase/dbstore.py:8:1: F401 'collections.defaultdict' imported but unused
./infogami/infobase/dbstore.py:12:1: F401 '_dbstore.schema.Schema' imported but unused
./infogami/infobase/dbstore.py:12:1: F401 '_dbstore.schema.INDEXED_DATATYPES' imported but unused
./infogami/infobase/common.py:1:1: F401 'config' imported but unused
./infogami/infobase/account.py:7:1: F401 'simplejson' imported but unused
./infogami/infobase/_dbstore/read.py:3:1: F401 'collections.defaultdict' imported but unused
./infogami/infobase/tests/test_store.py:6:1: F401 'simplejson' imported but unused
./infogami/infobase/tests/conftest.py:2:1: F401 'pytest_wildcard.*' imported but unused
./infogami/infobase/tests/test_seq.py:4:1: F401 'unittest' imported but unused
./infogami/infobase/tests/test_seq.py:5:1: F401 'simplejson' imported but unused
./infogami/infobase/tests/test_client.py:3:1: F401 'infogami.infobase.server' imported but unused
./infogami/utils/types.py:13:5: F401 'web' imported but unused
./infogami/utils/app.py:5:1: F401 're' imported but unused
./tests/test_doctests.py:2:1: F401 'doctest' imported but unused
25    F401 'web' imported but unused
```

File: infogami/utils/storage.py (hunk 8 above)
Comment (cclauss) [reply infogami/utils/storage.py]: Why is it more logical?

Comment (tfmorris) [reply infogami/utils/storage.py]: All the standard methods work on the merged entries from all the dicts. The only dict-specific method is add_dict. This currently returns the number of dicts, not the number of visible merged entries in the union of the dicts, as I'd expect.

File: infogami/infobase/__init__.py (hunk 4 above)
Comment (hornc) [reply infogami/infobase/__init__.py]: resolving per the flake8 evidence above not showing these as unsed..

File: infogami/utils/storage.py (hunk 8 above)
Comment (hornc) [reply infogami/utils/storage.py]: done, thanks for the review @tfmorris

**Comment by cclauss** [comment]
@hornc Tons of merge conflicts -- Do we resolve them or close this?

File: infogami/utils/app.py [hunk 10]
```diff
@@ -7,8 +7,7 @@
 
 import web
 
-import flash
-import delegate as infogami_delegate
```
Comment (hornc) [inline infogami/utils/app.py]: This is actually used further down and should not have been removed. https://github.com/infogami/infogami/issues/4 needs to be solved more directly.

Comment (cclauss) [reply infogami/utils/app.py]: __import delegate as infogami_delegate  # noqa__

**Comment by hornc** [comment]
This will not get merged. Was expoloratory only.

Some of the test fix investigation may be useful to pull across to master, so keep it around for reference. https://github.com/internetarchive/infogami/issues/23 needs to be done first.

**Pull request closed without merging**